File: superstitioApi/java/superstitioapi/cardModifier/RenderStackedBlockInstancesPatch.java
package superstitioapi.cardModifier;

import basemod.ClickableUIElement;
import basemod.ReflectionHacks;
import basemod.helpers.TooltipInfo;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.evacipated.cardcrawl.mod.stslib.blockmods.BlockInstance;
import com.evacipated.cardcrawl.mod.stslib.blockmods.BlockModifierManager;
import com.evacipated.cardcrawl.mod.stslib.patches.RenderStackedBlockInstances;
import com.evacipated.cardcrawl.modthespire.lib.SpireField;
import com.evacipated.cardcrawl.modthespire.lib.SpirePatch;
import com.evacipated.cardcrawl.modthespire.lib.SpirePrefixPatch;
import com.evacipated.cardcrawl.modthespire.lib.SpireReturn;
import com.megacrit.cardcrawl.core.AbstractCreature;
import com.megacrit.cardcrawl.core.Settings;
import com.megacrit.cardcrawl.helpers.FontHelper;
import com.megacrit.cardcrawl.helpers.PowerTip;
import com.megacrit.cardcrawl.helpers.TipHelper;
import javassist.CannotCompileException;
import javassist.expr.ExprEditor;
import javassist.expr.FieldAccess;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import static superstitioapi.cardModifier.RenderStackedBlockInstancesPatch.BlockStackForceShowPatch.GetBlockStack;

public class RenderStackedBlockInstancesPatch {
    protected static final float BLOCK_ICON_X = ReflectionHacks.getPrivateStatic(AbstractCreature.class, "BLOCK_ICON_X");
    protected static final float BLOCK_ICON_Y = ReflectionHacks.getPrivateStatic(AbstractCreature.class, "BLOCK_ICON_Y");
    private static final Texture blankTex = new Texture("images/blank.png");
    private static final float dx = 50.0F * Settings.scale;
    private static final float dy = 54.0F * Settings.scale;

    private static List<BlockStackDividedElement> makeAllBlockModifierIntoDrawElement(AbstractCreature creature) {
        List<BlockInstance> instances = getAllBlockInstances(creature);
        List<BlockStackDividedElement> list = new ArrayList<>();
        for (BlockInstance blockInstance : instances) {
            BlockStackDividedElement blockElement = new BlockStackDividedElement(creature, blockInstance);
            list.add(blockElement);
        }
        Collections.reverse(list);
        return list;
    }

    private static ArrayList<BlockInstance> getAllBlockInstances(AbstractCreature creature) {
        ArrayList<BlockInstance> blockInstances = BlockModifierManager.blockInstances(creature);
        ArrayList<BlockInstance> newBlockInstances = new ArrayList<>();
        newBlockInstances.addAll(blockInstances);
        newBlockInstances.addAll(creature.powers.stream()
                .filter(power -> power instanceof RenderAsBlockPower)
                .map(power -> ((RenderAsBlockPower) power).getBlockInstance()).collect(Collectors.toList()));
        return newBlockInstances;
    }

    private static void DrawBlocks(AbstractCreature creature, SpriteBatch sb, float x, float y) {
        float blockOffset = ReflectionHacks.getPrivate(creature, AbstractCreature.class, "blockOffset");
        float blockScale = ReflectionHacks.getPrivate(creature, AbstractCreature.class, "blockScale");
        Color blockOutlineColor = ReflectionHacks.getPrivate(creature, AbstractCreature.class, "blockOutlineColor");
        Color blockTextColor = ReflectionHacks.getPrivate(creature, AbstractCreature.class, "blockTextColor");

//        List<BlockInstance> allBlockLastTime = BlockStackElementField.blockLastTime.get(creature);
        List<BlockInstance> allBlocks = getAllBlockInstances(creature);
        List<BlockStackDividedElement> elements = BlockStackElementField.element.get(creature);
        if (allBlocks.stream().allMatch(BlockInstance::defaultBlock))
            return;
        if (elements == null || elements.isEmpty()
                || allBlocks.size() != elements.size()
                || !elements.stream().map(element -> element.blockInstance).allMatch(allBlocks::contains)) {
            elements = makeAllBlockModifierIntoDrawElement(creature);
        }

        if (elements.isEmpty())
            return;

//        Collections.reverse(elements);
        int offsetY = 0;
        for (BlockStackDividedElement element : elements) {
            element.move(x + BLOCK_ICON_X - 32.0F, y + BLOCK_ICON_Y - 32.0F + blockOffset + offsetY);
            element.update();
            element.renderBlockIcon(sb, blockScale, blockOutlineColor, blockTextColor);
            offsetY = (int) ((float) offsetY + dy);
        }

        BlockStackElementField.element.set(creature, elements);
    }

    public static class BlockStackDividedElement extends ClickableUIElement {
        public static final float HOVERED_TIMER_INIT = 1.0f;
        public static final float HOVERED_TIMER_MIN = 0.2f;
        private static final float baseHeight = 64.0F * Settings.scale;
        private static final float baseWidth = 64.0F * Settings.scale;
        private final AbstractCreature owner;
        private final BlockInstance blockInstance;
        private PowerTip tip;
        private float HoveredTimer = HOVERED_TIMER_INIT;

        public BlockStackDividedElement(AbstractCreature owner, BlockInstance blockInstance) {
            super(blankTex, 0.0F, 0.0F, baseWidth, baseHeight);
            this.owner = owner;
            this.blockInstance = blockInstance;
//            this.tips.addAll(Arrays.asList(tips));
        }

        public void move(float x, float y) {
            this.move(x, y, 0.0F);
        }

        private void move(float x, float y, float d) {
            this.setX(x - d * Settings.scale);
            this.setY(y - d * Settings.scale);
        }

        public void setHitboxHeight(float height) {
            this.hitbox.resize(baseWidth, baseHeight + height);
        }

        public void renderBlockIcon(SpriteBatch sb, float blockScale, Color blockOutlineColor, Color blockTextColor) {
            this.render(sb);

            Color IconColor = blockInstance.getBlockColor() != null ? blockInstance.getBlockColor() : blockOutlineColor;
            Color TextColor = blockInstance.getBlockColor() != null ? blockInstance.getBlockColor() : blockTextColor;

            float tmpIcon = IconColor.a;
            float tmpText = TextColor.a;
            if (hitbox.hovered) {
                IconColor.a *= HoveredTimer;
                TextColor.a *= HoveredTimer;
            }

            sb.setColor(IconColor);


            sb.draw(blockInstance.getBlockImage(), x, y,
                    32.0F, 32.0F, 64.0F, 64.0F, Settings.scale, Settings.scale, 0.0F, 0, 0, 64, 64, false, false);

            FontHelper.renderFontCentered(sb, FontHelper.blockInfoFont, Integer.toString(getBlockAmount()),
                    x - (-32.0F), y - (BLOCK_ICON_Y - 32.0F) - 16.0F * Settings.scale, TextColor, blockScale);

            IconColor.a = tmpIcon;
            TextColor.a = tmpText;

            if (hitbox.hovered)
                FontHelper.renderFontCentered(sb, FontHelper.blockInfoFont, Integer.toString(GetBlockStack(owner)),
                        x - (-32.0F), y - (BLOCK_ICON_Y - 32.0F) - 16.0F * Settings.scale, TextColor, blockScale);

        }

        private int getBlockAmount() {
            return blockInstance.getBlockAmount();
        }

        @Override
        public void update() {
            super.update();
        }

        @Override
        protected void onHover() {
            if (HoveredTimer >= HOVERED_TIMER_MIN) {
                HoveredTimer -= Gdx.graphics.getDeltaTime();
            }
            ArrayList<PowerTip> tips = new ArrayList<>();
            tips.add(new TooltipInfo(blockInstance.makeName(), blockInstance.makeDescription()).toPowerTip());
            boolean isInScreen = this.hitbox.width / 2.0F < 1544.0F * Settings.scale;
            TipHelper.queuePowerTips(
                    isInScreen ? this.owner.hb.cX + this.owner.hb.width / 2.0F + 20.0F * Settings.scale
                            : this.owner.hb.cX - this.owner.hb.width / 2.0F + -380.0F * Settings.scale,
                    this.y + this.owner.hb.height + TipHelper.calculateAdditionalOffset(tips, this.y + this.owner.hb.height), tips);
        }

        @Override
        protected void onUnhover() {
            if (HoveredTimer < HOVERED_TIMER_INIT)
                HoveredTimer = Math.min(HoveredTimer + Gdx.graphics.getDeltaTime(), HOVERED_TIMER_INIT);
        }

        @Override
        protected void onClick() {
        }
    }

    @SpirePatch(
            clz = AbstractCreature.class,
            method = "<class>"
    )
    public static class BlockStackElementField {
        public static SpireField<List<BlockStackDividedElement>> element = new SpireField<>(() -> null);
        public static SpireField<Boolean> forceDrawBlock = new SpireField<>(() -> false);
    }

    @SpirePatch(clz = RenderStackedBlockInstances.RenderStackedIcons.class, method = "pls")
    public static class BlockStackPatchStslib {
        public static SpireReturn<Void> Prefix(AbstractCreature __instance, SpriteBatch sb, float x, float y, float ___BLOCK_ICON_X,
                                               float ___BLOCK_ICON_Y, float ___blockOffset, Color ___blockTextColor, Color ___blockOutlineColor,
                                               float ___blockScale) {
            DrawBlocks(__instance, sb, x, y);
            return SpireReturn.Return();
        }
    }

    @SpirePatch(
            clz = AbstractCreature.class,
            method = "renderBlockIconAndValue"
    )
    public static class BlockStackPatchVanilla {
        @SpirePrefixPatch
        public static SpireReturn<Void> Prefix(AbstractCreature __instance, SpriteBatch sb, float x, float y) {
            if (getAllBlockInstances(__instance).stream().allMatch(BlockInstance::defaultBlock))
                return SpireReturn.Continue();
            return SpireReturn.Return();
        }
    }

    @SpirePatch(
            clz = AbstractCreature.class,
            method = "renderHealth"
    )
    public static class BlockStackForceShowPatch {
        public static ExprEditor Instrument() {
            return new ExprEditor() {
                public void edit(final FieldAccess m) throws CannotCompileException {
                    if (!m.getFieldName().equals("currentBlock") || !m.isReader()) return;
                    m.replace("{$_ = " + BlockStackForceShowPatch.class.getName() + ".GetBlockStack( $0 ) ;}");


                }
            };
        }

        public static int GetBlockStack(AbstractCreature creature) {
            if (!BlockStackElementField.forceDrawBlock.get(creature)) return creature.currentBlock;
            int totalBlock = RenderStackedBlockInstancesPatch.getAllBlockInstances(creature).stream()
                    .mapToInt(BlockInstance::getBlockAmount).sum();
            if (totalBlock <= 0) {
                BlockStackElementField.forceDrawBlock.set(creature, false);
                return creature.currentBlock;
            }
            return totalBlock;
        }
    }

}

File: superstitioApi/java/superstitioapi/cardModifier/RenderStackedBlockInstancesPatch.kt
package superstitioapi.cardModifier

import basemod.ClickableUIElement
import basemod.ReflectionHacks
import basemod.helpers.TooltipInfo
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.evacipated.cardcrawl.mod.stslib.blockmods.BlockInstance
import com.evacipated.cardcrawl.mod.stslib.blockmods.BlockModifierManager
import com.evacipated.cardcrawl.mod.stslib.patches.RenderStackedBlockInstances.RenderStackedIcons
import com.evacipated.cardcrawl.modthespire.lib.*
import com.megacrit.cardcrawl.core.AbstractCreature
import com.megacrit.cardcrawl.core.Settings
import com.megacrit.cardcrawl.helpers.FontHelper
import com.megacrit.cardcrawl.helpers.PowerTip
import com.megacrit.cardcrawl.helpers.TipHelper
import com.megacrit.cardcrawl.powers.AbstractPower
import javassist.CannotCompileException
import javassist.expr.ExprEditor
import javassist.expr.FieldAccess
import java.util.*
import kotlin.math.min

/**
 * 此类用于修改游戏中的渲染逻辑，以更准确地显示堆叠的格挡效果。
 * 它修改了如何显示生物的格挡数量，确保所有格挡修饰符都被考虑在内。
 */
object RenderStackedBlockInstancesPatch
{
    // 常量定义，用于确定格挡图标的位置和大小
    internal val BLOCK_ICON_X: Float = ReflectionHacks.getPrivateStatic(AbstractCreature::class.java, "BLOCK_ICON_X")
    internal val BLOCK_ICON_Y: Float = ReflectionHacks.getPrivateStatic(AbstractCreature::class.java, "BLOCK_ICON_Y")
    private val blankTex = Texture("images/blank.png")
    private val dx = 50.0f * Settings.scale
    private val dy = 54.0f * Settings.scale

    /**
     * 生成一个列表，其中包含给定生物的所有格挡修饰符对应的绘制元素。
     * 这些元素稍后将用于渲染格挡图标。
     *
     * @param creature 给定的生物
     * @return 包含BlockStackDividedElement的列表，每个元素代表一个格挡实例
     */
    private fun makeAllBlockModifierIntoDrawElement(creature: AbstractCreature): List<BlockStackDividedElement>
    {
        // 获取所有格挡实例并创建对应的绘制元素
        val instances: List<BlockInstance?> = getAllBlockInstances(creature)
        val list: MutableList<BlockStackDividedElement> = ArrayList()
        for (blockInstance in instances)
        {
            val blockElement = BlockStackDividedElement(creature, blockInstance)
            list.add(blockElement)
        }
        list.reverse() // 反转列表以确保正确的渲染顺序
        return list
    }

    /**
     * 获取给定生物的所有格挡实例，包括那些由可以渲染为格挡效果的力量产生的实例。
     *
     * @param creature 目标生物
     * @return 包含所有格挡效果的BlockInstance列表
     */
    private fun getAllBlockInstances(creature: AbstractCreature): MutableList<BlockInstance>
    {
        // 合并生物的格挡实例和可以渲染为格挡效果的Power
        val blockInstances = BlockModifierManager.blockInstances(creature)
        val newBlockInstances = ArrayList<BlockInstance>()
        newBlockInstances.addAll(blockInstances)
        creature.powers
            .filter { power: AbstractPower? -> power is RenderAsBlockPower }
            .map { power: AbstractPower -> (power as RenderAsBlockPower).getBlockInstance() }
            .let(newBlockInstances::addAll)
        return newBlockInstances
    }

    /**
     * 渲染给定生物的格挡图标，考虑到所有的格挡实例和修饰符。
     *
     * @param creature 目标生物
     * @param sb       用于渲染的SpriteBatch对象
     * @param x        开始渲染的x坐标
     * @param y        开始渲染的y坐标
     */
    private fun DrawBlocks(creature: AbstractCreature, sb: SpriteBatch, x: Float, y: Float)
    {
        // 获取渲染格挡所需的设置和颜色
        val blockOffset = ReflectionHacks.getPrivate<Float>(creature, AbstractCreature::class.java, "blockOffset")
        val blockScale = ReflectionHacks.getPrivate<Float>(creature, AbstractCreature::class.java, "blockScale")
        val blockOutlineColor =
            ReflectionHacks.getPrivate<Color>(creature, AbstractCreature::class.java, "blockOutlineColor")
        val blockTextColor = ReflectionHacks.getPrivate<Color>(creature, AbstractCreature::class.java, "blockTextColor")

        // 检查是否需要重新生成绘制元素
        val allBlocks: List<BlockInstance?> = getAllBlockInstances(creature)
        var elements: List<BlockStackDividedElement>? = BlockStackElementField.element[creature]
        if (allBlocks.stream().allMatch { obj: BlockInstance? -> obj!!.defaultBlock() }) return
        if (elements == null || elements.isEmpty() || allBlocks.size != elements.size || !elements.stream()
                .map { element: BlockStackDividedElement? -> element!!.blockInstance }
                .allMatch(allBlocks::contains)
        )
        {
            elements = makeAllBlockModifierIntoDrawElement(creature)
        }

        if (elements.isEmpty()) return

        // 遍历并渲染每个格挡元素
        var offsetY = 0
        for (element in elements)
        {
            element.move(x + BLOCK_ICON_X - 32.0f, y + BLOCK_ICON_Y - 32.0f + blockOffset + offsetY)
            element.update()
            element.renderBlockIcon(sb, blockScale, blockOutlineColor, blockTextColor)
            offsetY = (offsetY.toFloat() + dy).toInt()
        }

        // 更新生物的格挡元素字段
        BlockStackElementField.element.set(creature, elements)
    }

    class BlockStackDividedElement //            this.tips.addAll(Arrays.asList(tips));
        (private val owner: AbstractCreature, val blockInstance: BlockInstance?) :
        ClickableUIElement(blankTex, 0.0f, 0.0f, baseWidth, baseHeight)
    {
        private val tip: PowerTip? = null
        private var HoveredTimer = HOVERED_TIMER_INIT

        fun move(x: Float, y: Float)
        {
            this.move(x, y, 0.0f)
        }

        fun setHitboxHeight(height: Float)
        {
            hitbox.resize(baseWidth, baseHeight + height)
        }

        fun renderBlockIcon(sb: SpriteBatch, blockScale: Float, blockOutlineColor: Color?, blockTextColor: Color?)
        {
            this.render(sb)

            val IconColor = if (blockInstance!!.blockColor != null) blockInstance.blockColor else blockOutlineColor!!
            val TextColor = if (blockInstance.blockColor != null) blockInstance.blockColor else blockTextColor!!

            val tmpIcon = IconColor.a
            val tmpText = TextColor.a
            if (hitbox.hovered)
            {
                IconColor.a *= HoveredTimer
                TextColor.a *= HoveredTimer
            }

            sb.color = IconColor


            sb.draw(
                blockInstance.blockImage, x, y,
                32.0f, 32.0f, 64.0f, 64.0f, Settings.scale, Settings.scale, 0.0f, 0, 0, 64, 64, false, false
            )

            FontHelper.renderFontCentered(
                sb, FontHelper.blockInfoFont, blockAmount.toString(),
                x - (-32.0f), y - (BLOCK_ICON_Y - 32.0f) - 16.0f * Settings.scale, TextColor, blockScale
            )

            IconColor.a = tmpIcon
            TextColor.a = tmpText

            if (hitbox.hovered) FontHelper.renderFontCentered(
                sb, FontHelper.blockInfoFont, BlockStackForceShowPatch.GetBlockStack(
                    owner
                ).toString(),
                x - (-32.0f), y - (BLOCK_ICON_Y - 32.0f) - 16.0f * Settings.scale, TextColor, blockScale
            )
        }

        private fun move(x: Float, y: Float, d: Float)
        {
            this.setX(x - d * Settings.scale)
            this.setY(y - d * Settings.scale)
        }

        private val blockAmount: Int
            get() = blockInstance!!.blockAmount

        override fun onHover()
        {
            if (HoveredTimer >= HOVERED_TIMER_MIN)
            {
                HoveredTimer -= Gdx.graphics.deltaTime
            }
            val tips = ArrayList<PowerTip>()
            tips.add(TooltipInfo(blockInstance!!.makeName(), blockInstance.makeDescription()).toPowerTip())
            val isInScreen = hitbox.width / 2.0f < 1544.0f * Settings.scale
            TipHelper.queuePowerTips(
                if (isInScreen) owner.hb.cX + (owner.hb.width / 2.0f) + (20.0f * Settings.scale)
                else owner.hb.cX - owner.hb.width / 2.0f + -380.0f * Settings.scale,
                this.y + owner.hb.height + TipHelper.calculateAdditionalOffset(tips, this.y + owner.hb.height), tips
            )
        }

        override fun onUnhover()
        {
            if (HoveredTimer < HOVERED_TIMER_INIT) HoveredTimer = min(
                (HoveredTimer + Gdx.graphics.deltaTime).toDouble(),
                HOVERED_TIMER_INIT.toDouble()
            ).toFloat()
        }

        override fun onClick()
        {
        }

        companion object
        {
            const val HOVERED_TIMER_INIT: Float = 1.0f
            const val HOVERED_TIMER_MIN: Float = 0.2f
            private val baseHeight = 64.0f * Settings.scale
            private val baseWidth = 64.0f * Settings.scale
        }
    }

    @SpirePatch2(clz = AbstractCreature::class, method = "<class>")
    object BlockStackElementField
    {
        @JvmField
        var element: SpireField<List<BlockStackDividedElement>?> = SpireField<List<BlockStackDividedElement>?> { null }

        @JvmField
        var forceDrawBlock: SpireField<Boolean> = SpireField<Boolean> { false }
    }

    @SpirePatch(clz = RenderStackedIcons::class, method = "pls")
    object BlockStackPatchStslib
    {
        @SpirePrefixPatch
        @JvmStatic
        fun Prefix(
            __instance: AbstractCreature, sb: SpriteBatch, x: Float, y: Float, ___BLOCK_ICON_X: Float,
            ___BLOCK_ICON_Y: Float, ___blockOffset: Float, ___blockTextColor: Color?, ___blockOutlineColor: Color?,
            ___blockScale: Float
        ): SpireReturn<Void>
        {
            DrawBlocks(__instance, sb, x, y)
            return SpireReturn.Return()
        }
    }

    @SpirePatch2(clz = AbstractCreature::class, method = "renderBlockIconAndValue")
    object BlockStackPatchVanilla
    {
        @SpirePrefixPatch
        @JvmStatic
        fun Prefix(__instance: AbstractCreature, sb: SpriteBatch?, x: Float, y: Float): SpireReturn<Void>
        {
            if (getAllBlockInstances(__instance).stream()
                    .allMatch { obj: BlockInstance? -> obj!!.defaultBlock() }
            ) return SpireReturn.Continue()
            return SpireReturn.Return()
        }
    }

    @SpirePatch2(clz = AbstractCreature::class, method = "renderHealth")
    object BlockStackForceShowPatch
    {
        @SpireInstrumentPatch
        @JvmStatic
        fun Instrument(): ExprEditor
        {
            return object : ExprEditor()
            {
                @Throws(CannotCompileException::class)
                override fun edit(m: FieldAccess)
                {
                    if (m.fieldName != "currentBlock" || !m.isReader) return
                    m.replace("{\$_ = " + BlockStackForceShowPatch::class.qualifiedName + ".GetBlockStack( $0 ) ;}")
                }
            }
        }

        @JvmStatic
        fun GetBlockStack(creature: AbstractCreature): Int
        {
            if (!BlockStackElementField.forceDrawBlock[creature]) return creature.currentBlock
            val totalBlock = getAllBlockInstances(creature).stream()
                .mapToInt { obj: BlockInstance? -> obj!!.blockAmount }.sum()
            if (totalBlock <= 0)
            {
                BlockStackElementField.forceDrawBlock[creature] = false
                return creature.currentBlock
            }
            return totalBlock
        }
    }
}
